i6_models/config.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import MISSING, dataclass
+from typing import Any, List, Optional
+
+
+@dataclass
+class ModelConfiguration:
+    """base dataclass that supported fetching attributes"""
+
+    _name: Optional[str] = None
+
+    @staticmethod
+    def name() -> Optional[str]:
+        return None
+
+    def _get_all_attributes(self) -> List[str]:
+        return [k for k in self.__dataclass_fields__.keys()]
+
+    def _get_name(self, attribute_name: str) -> str:
+        return self.__dataclass_fields__[attribute_name].name
+
+    def _get_default(self, attribute_name: str) -> Any:
+        if hasattr(self, attribute_name):
+            return getattr(self, attribute_name)
+
+        f = self.__dataclass_fields__[attribute_name]
+        if f.default_factory is not MISSING:
+            return f.default_factory()
+        return f.default
+
+    def _get_type(self, attribute_name: str) -> Any:
```

Comment:
Where is this function used?

Reply:
I don't understand. This really depends on your usecase. Here this is just abstract implementation for use in derived configs. Same goes for similar comments.

Reply:
I don't see any other code which uses this. So if some code is not used, I would remove it. We should not have code here which might be used later by other code but is not used yet.

Reply:
This class is supposed to offer basic functionality for configs and while you are right that we should carefully decide which functions are helpful and which are not, I think `_get_type` is something that will be used. Similar to the other 3 private functions. Regarding the name, maybe we should remove that for now, I am not sure.

Reply:
Ok, but then my question is, where will it be used, how will it be used? I don't really see any documentation here, nor any usage, so it's hard to just guess the potential usage. It would help to see some actual code where it is used, and maybe also some documentation.

As this is a private method, I assume the usage will only be internal. So what internal code would use it? Why did you not include this basic functionality?

Reply:
An example would be to check for proper initialization of the config. E.g. you would want `num_layers` to always be int. So instead of checking for each parameter individually you can check whether the value the user gives is the same type as the field.

Reply:
Who do you mean by "you"? The user? I thought this is only for internal use, i.e. the user would never use it. It the user is supposed to use it, it should not be marked as private, i.e. not start with underscore.

Is type checking supposed to be done anyway automatically? I would have expected that all types would always be checked, or not? But where is that code?

Reply:
>Is type checking supposed to be done anyway automatically?

No. This was just some example I can up with.

>I thought this is only for internal use

Yes. From this e.g. `ConformerConfig` is derived where then `_get_type` would be used (internally). Type checking was just an example I came up with.

Reply:
I think the major discussion point here is, which functions this base class should provide and which are deemed not helpful for now. Maybe we can discuss this in todays meeting, or other can comment? 
I don't really have a strong opinion on those function, those where just the ones I thought would be helpful looking at the Fairseq class.

Reply:
So, using the example:
```python
@dataclass
class ExampleConfig(ModelConfiguration):
    hidden_dim: int = 256
    name: str = "Example Configuration"

class ExampleModule(Module):
    def __init__(self, cfg: ExampleConfig)
        self.hidden_dim = cfg.hidden_dim
```
Now when the user writes:
```python
mod = ExampleModule(ExampleConfig(hidden_dim=2.5))
```
I thought that some code here would check the types and raise an exception here that `hidden_dim` is wrong. But this would be handled by `ModelConfiguration` automatically for all attribs. This is why I thought we have this `_get_type` method. Otherwise I still don't really understand why we need the `_get_type` method.

Or do you think that in `ExampleModule`, there would be an extra check like:
```python
assert isinstance(cfg.hidden_dim, cfg._get_type("hidden_dim"))
```
But this would be bad, as you would access a private method from outside. Also, this is not really nice to read. In this example, I would simply prefer:
```python
assert isinstance(cfg.hidden_dim, int)
```
Even if you would replicate the type `int` both here and in the `ExampleConfig` definition.

Or is this not the purpose of the function? I still don't really see the purpose. The only purpose which makes sense for me is if this would be used by some automatic type checking for all attribs. But that's why I was asking for some examples of potential usages.

Reply:
> do you mean we should add code to the config class to do type checking automatically?

I think it's a good idea to force correct typing in the config and throw an exception if this is violated.

Reply:
I was also wondering whether `dataclass` itself already supports such a functionality. Maybe by defining `_get_type`, this is actually the way how `dataclass` would handle it? I don't know. That's why I was asking here.

Reply:
[Relevant StackOverflow question](https://stackoverflow.com/questions/50563546/validating-detailed-types-in-python-dataclasses).

Reply:
a quick search brought me to this library that enables type checking at runtime with a decorator 
https://pypi.org/project/typeguard/

we could just add this to the dataclass and all should be checked automatically

Reply:
> a quick search brought me to this library that enables type checking at runtime with a decorator https://pypi.org/project/typeguard/
> 
> we could just add this to the dataclass and all should be checked automatically

Are you sure this works here?

And this would not need the `_get_type` method here, right?

Reply:
I am currently trying to set `typeguard` up, but I think it is not that straight forward.

Reply:
ESPnet actually wraps the `check_type` function. 
https://espnet.github.io/espnet/_modules/espnet2/utils/build_dataclass.html

Reply:
See my StackOverflow link above. It is specifically about possible solutions for `dataclass`.